src/dynamic_honey_badger.rs

```diff
@@ -0,0 +1,450 @@
+        &mut self,
+        sender_id: &NodeUid,
+        propose: Propose,
+        sig: &Signature,
+    ) -> Result<()> {
+        if !self.verify_signature(sender_id, sig, &propose)? {
+            debug!("Invalid signature from {:?} for: {:?}.", sender_id, propose);
+            return Ok(());
+        }
+        let handle =
+            |key_gen: &mut SyncKeyGen<NodeUid>| key_gen.handle_propose(&sender_id, propose);
+        let accept = match self.key_gen.as_mut().and_then(handle) {
+            Some(accept) => accept,
+            None => return Ok(()),
+        };
+        let our_id = self.our_id().clone();
+        let sig = self.sign(&accept)?;
+        let tx = Transaction::Accept(our_id, accept, sig);
+        self.honey_badger.input(tx)?;
+        Ok(())
+    }
+
+    /// Handles an `Accept` message that was output by Honey Badger.
+    fn handle_accept(
+        &mut self,
+        sender_id: &NodeUid,
+        accept: Accept,
+        sig: &Signature,
+    ) -> Result<()> {
+        if self.verify_signature(sender_id, sig, &accept)? {
+            if let Some(key_gen) = self.key_gen.as_mut() {
+                key_gen.handle_accept(&sender_id, accept);
+            }
+        }
+        Ok(())
+    }
+
+    /// If the current Key Generation process is ready, returns the generated key set.
+    fn get_key_gen_output(&self) -> Option<(PublicKeySet, ClearOnDrop<Box<SecretKey>>)> {
+        // TODO: Once we've upgraded to Rust 1.27.0, we can use `Option::filter` here.
+        self.key_gen
+            .iter()
+            .filter(|key_gen| key_gen.is_ready())
+            .map(SyncKeyGen::generate)
+            .next()
+    }
+
+    /// Returns a signature of `payload`, or an error if serialization fails.
+    fn sign<T: Serialize>(&self, payload: &T) -> Result<Box<Signature>> {
+        let ser = bincode::serialize(payload)?;
+        Ok(Box::new(self.netinfo.secret_key().sign(ser)))
+    }
+
+    /// Returns `true` if the signature of the payload by the node with the specified ID is valid.
+    /// Returns an error if the payload fails to serialize.
+    fn verify_signature<T: Serialize>(
+        &self,
+        id: &NodeUid,
+        sig: &Signature,
+        payload: &T,
+    ) -> Result<bool> {
+        let ser = bincode::serialize(payload)?;
+        let pk_opt = self.netinfo.public_key_share(&id);
+        Ok(pk_opt.map_or(false, |pk| pk.verify(&sig, ser)))
+    }
+
+    /// Adds a vote for a node change by the node with `id`.
+    fn insert_vote(&mut self, id: NodeUid, change: Change<NodeUid>) {
+        if let Some(old_change) = self.votes.insert(id, change.clone()) {
+            let decrement = |count: &mut usize| {
+                *count -= 1;
+                *count
+            };
+            if Some(0) == self.vote_counts.get_mut(&old_change).map(decrement) {
+                self.vote_counts.remove(&old_change);
+            }
+        }
+        *self.vote_counts.entry(change).or_insert(0) += 1;
+    }
```

Comment:
Would the following code be equivalent to `insert_vote` and yet shorter?
```
if self.votes.insert(id, change.clone()).is_none() {
    *self.vote_counts.entry(change).or_insert(0) += 1;
}
```

Reply:
Not quite. Sorry for my lousy code comments: The idea is that every node can only have one active vote at a time; voting again revokes the previous vote.
So when we insert a vote, we need to increment the count for that vote, and also decrement the count for the _previous_ vote of that node. Not sure how that can be expressed more concisely.

Reply:
One way would be to avoid duplication by removing `vote_counts` but it does look like a handy optimization. Maybe you could mention this semantics of revoking votes in a comment to `Change` so that it is visible?

Reply:
No, you're right, removing it is better: the optimization is pointless since it's only computed twice per epoch. Also it required more cloning, and if there are a lot of votes it might even be _slower_, so I removed it.